hop/io.py

```diff
@@ -3,44 +3,218 @@
+            in write mode or an adc Consumer instance in read mode.
+
+        Raises:
+            ValueError: If the mode is not set to read/write or if more than
+                one topic is specified in write mode.
+
+        """
+        group_id, broker_addresses, topics = kafka.parse_kafka_url(url)
+        logger.debug("connecting to addresses=%s  group_id=%s  topics=%s",
+                     broker_addresses, group_id, topics)
+        if not auth and self.auth:
+            auth = self.auth
+        if mode == "w":
+            if len(topics) != 1:
+                raise ValueError("must specify exactly one topic in write mode")
+            if group_id is not None:
+                warnings.warn("group ID has no effect when opening a stream in write mode")
+            if start_at is not None:
+                warnings.warn("start_at has no effect when opening a stream in write mode")
+            if persist is not None:
+                warnings.warn("read_forever has no effect when opening a stream in write mode")
+            return _open_producer(broker_addresses, topics[0], auth=auth)
+        elif mode == "r":
+            if group_id is None:
+                group_id = _generate_group_id(10)
+                logger.info("group ID not specified, generating a random group ID")
+            # set up extra options if provided
+            opts = {}
+            if start_at or self.start_at:
+                opts["start_at"] = start_at if start_at else self.start_at
+            if persist is not None or self.persist is not None:
+                opts["read_forever"] = persist if persist is not None else self.persist
+            return _open_consumer(
+                group_id,
+                broker_addresses,
+                topics,
+                auth=auth,
+                metadata=metadata,
+                **opts,
+            )
+        else:
+            raise ValueError("mode must be either 'w' or 'r'")
+
+
+class Deserializer(Enum):
+    CIRCULAR = models.GCNCircular
+    VOEVENT = models.VOEvent
+    BLOB = models.MessageBlob
+
+    @classmethod
+    def deserialize(cls, message):
+        """Deserialize a stream message and instantiate a model.
+
+        Args:
+            message: A serialized message.
+
+        Returns:
+            A data container corresponding to the format in the
+            serialized message.
+
+        Raises:
+            ValueError: If the message is incorrectly formatted or
+                if the message format is not recognized.
 
         """
-        opts = {**self._options, **kwargs}
-        return streaming.open(*args, **opts)
+        try:
+            format = message["format"].upper()
+            content = message["content"]
+        except (KeyError, TypeError):
+            raise ValueError("Message is incorrectly formatted")
+        else:
+            if format in cls.__members__:
+                return cls[format.upper()].value(**content)
```

Comment:
I just noticed this `**content` bit. I think that means that we're expecting `content` to always be a dictionary. What if it's arbitrary bytes?

For example, what if it's something like base64-encoded Avro?

```
{
    "format": "LSST_ALERT_AVRO",
    "content": "AAAAAADQ3A2St+vd2gHAsIkBAAAAZCPYoNLS7EAGdGVkAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYHeXSQAA+EIAAPhCABxCRgAcQk",
}
```

Maybe we're saying that `content` **must** be a JSON dictionary? That works, but we should be clear about it I think.

Reply:
That is definitely a bug. All the unstructured messages I was testing with were JSON dictionaries which may explain why I didn't catch it. I definitely want to support any arbitrary bytes, but would have to think about how to make that work here, since that would fail out in the current implementation.

Reply:
So, I have something working that should support arbitrary content as long as it's JSON serializable. That doesn't include bytes, but does include base64 encoded strings. I think if we want to support raw bytes we'd have to auto-detect it and decode/encode as needed or move to Avro which I believe has a native bytes type.

Reply:
Right, if it's truly raw bytes, I don't think it we'd have it anyway: we wouldn't be able to `json.loads` the message payload, so we wouldn't have the `message["content"]` anyway. I think its fair to require base64-encoding as long as we have a JSON wrapper.